The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7291486075/job/19870515892) [(plain)](https://github.com/rust-lang/rust/commit/5adf217d42795c92eee059b7adabb664072b1eb5/checks/19870515892/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=asquared31415
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_48779fd7-116d-49d1-b7bd-203064a42d33
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=e1d0abad07cd96452a37c29f950a64287773878d
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_48779fd7-116d-49d1-b7bd-203064a42d33
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_48779fd7-116d-49d1-b7bd-203064a42d33
GITHUB_TRIGGERING_ACTOR=asquared31415
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/119195/merge
GITHUB_WORKFLOW_SHA=e1d0abad07cd96452a37c29f950a64287773878d
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Built container sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Looks like docker image is the same as before, not uploading
https://ci-caches.rust-lang.org/docker/7ebc15c01a233894034d277c8cce4e949f4e7791f66b4727c8fb6e058a0b8171d6152e1441d677cef0653843ceeee469c097b8699b2bb74249e674f6aa1a8813
sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Thu Dec 21 17:47:40 UTC 2023
  network time: Thu, 21 Dec 2023 17:47:40 GMT
  network time: Thu, 21 Dec 2023 17:47:40 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-missing-tools', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: rust.codegen-units-std := 1
---

---- [ui] tests/ui/asm/named-asm-labels.rs stdout ----
diff of stderr:

- error: avoid using named labels in inline assembly
+ error: invalid asm template modifier for this register class
+   --> $DIR/named-asm-labels.rs:134:20
3    |
3    |
- LL |         asm!("bar: nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
-    = note: `#[deny(named_asm_labels)]` on by default
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("abcd:");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("foo: bar1: nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("foo1: nop", "nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("foo2: foo3: nop", "nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("nop", "foo4: nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("foo5: nop", "foo6: nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("foo5: nop", "foo6: nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("foo7: nop; foo8: nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("foo9: nop; nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("nop; foo10: nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("bar2: nop\n bar3: nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("bar4: nop\n nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("nop\n bar5: nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("nop\n bar6: bar7: nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |             blah2: nop
-    |             ^^^^^
- LL |             blah3: nop
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |             nop ; blah4: nop
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("blah1: 2bar: nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("def: def: nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("def: nop\ndef: nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("def: nop; def: nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("fooo\u{003A} nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("foooo\x3A nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("fooooo:\u{000A} nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("foooooo:\x0A nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("\x41\x42\x43\x3A\x20\x6E\x6F\x70");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |             ab: nop // ab: does foo
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |             asm!("test_{}: nop", in(reg) 10);
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
259 LL |         asm!("test_{0:w}: nop", in(reg) 10);
+    |                    ^^^^^        ---------- argument
+    |                    |
+    |                    template modifier
261    |
261    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
+    = note: the `reg` register class supports the following template modifiers: `l`, `x`, `e`, `r`
264 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("test_{}: nop", const 10);
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("test_{}: nop", sym main);
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("{}_test: nop", const 10);
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("test_{}_test: nop", const 10);
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("{}: nop", const 10);
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("{uwu}: nop", uwu = const 10);
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("{0}: nop", const 10);
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("{1}: nop", "/* {0} */", const 10, const 20);
-    |               ^^^
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!(include_str!("named-asm-labels.s"));
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- warning: avoid using named labels in inline assembly
-    |
-    |
- LL |             asm!("warned: nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
-   --> $DIR/named-asm-labels.rs:154:16
-    |
-    |
- LL |         #[warn(named_asm_labels)]
- 
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |     unsafe { asm!(".Lfoo: mov rax, {}; ret;", "nop", const 1, options(noreturn)) }
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |     unsafe { asm!(".Lbar: mov rax, {}; ret;", "nop", const 1, options(noreturn)) }
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |     unsafe { asm!(".Laaa: nop; ret;", options(noreturn)) }
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         unsafe { asm!(".Lbbb: nop; ret;", options(noreturn)) }
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("closure1: nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |         asm!("closure2: nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- 
- error: avoid using named labels in inline assembly
-    |
-    |
- LL |             asm!("closure3: nop");
-    |
-    |
-    = help: only local labels of the form `<number>:` should be used in inline asm
-    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
- error: aborting due to 45 previous errors; 1 warning emitted
+ error: aborting due to 1 previous error
424 
425 
---
--- stderr -------------------------------
error: invalid asm template modifier for this register class
##[error]  --> /checkout/tests/ui/asm/named-asm-labels.rs:134:20
   |
LL |         asm!("test_{0:w}: nop", in(reg) 10); //~ ERROR avoid using named labels
   |                    |
   |                    template modifier
   |
   |
   = note: the `reg` register class supports the following template modifiers: `l`, `x`, `e`, `r`
error: aborting due to 1 previous error
------------------------------------------


```

</details>

--- a/compiler/rustc_lint/src/builtin.rs
+++ b/compiler/rustc_lint/src/builtin.rs
@@ -2734,10 +2734,13 @@ impl<'tcx> LateLintPass<'tcx> for NamedAsmLabels {
     #[allow(rustc::diagnostic_outside_of_impl)]
     fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'tcx>) {
         if let hir::Expr {
-            kind: hir::ExprKind::InlineAsm(hir::InlineAsm { template_strs, .. }),
+            kind: hir::ExprKind::InlineAsm(hir::InlineAsm { template_strs, options, .. }),
             ..
         } = expr
         {
+            // asm with `options(raw)` does not do replacement with `{` and `}`.
+            let raw = options.contains(InlineAsmOptions::RAW);
+
             for (template_sym, template_snippet, template_span) in template_strs.iter() {
                 let template_str = template_sym.as_str();
                 let find_label_span = |needle: &str| -> Option<Span> {
@@ -2763,24 +2766,57 @@ impl<'tcx> LateLintPass<'tcx> for NamedAsmLabels {
                 for statement in statements {
                     // If there's a comment, trim it from the statement
                     let statement = statement.find("//").map_or(statement, |idx| &statement[..idx]);
+
+                    // In this loop, if there is ever a non-label, no labels can come after it.
                     let mut start_idx = 0;
-                    for (idx, _) in statement.match_indices(':') {
+                    'label_loop: for (idx, _) in statement.match_indices(':') {
                         let possible_label = statement[start_idx..idx].trim();
                         let mut chars = possible_label.chars();
-                        let Some(c) = chars.next() else {
-                            // Empty string means a leading ':' in this section, which is not a label
-                            break;
+
+                        let Some(start) = chars.next() else {
+                            // Empty string means a leading ':' in this section, which is not a label.
+                            break 'label_loop;
                         };
-                        // A label starts with an alphabetic character or . or _ and continues with alphanumeric characters, _, or $
-                        if (c.is_alphabetic() || matches!(c, '.' | '_'))
-                            && chars.all(|c| c.is_alphanumeric() || matches!(c, '_' | '$'))
-                        {
-                            found_labels.push(possible_label);
-                        } else {
-                            // If we encounter a non-label, there cannot be any further labels, so stop checking
-                            break;
+
+                        // Whether a { bracket has been seen and its } hasn't been found yet.
+                        let mut in_bracket = false;
+
+                        // A label starts with an ASCII alphabetic character or . or _
+                        // A label can also start with a format arg, if it's not a raw asm block.
+                        if !raw && start == '{' {
+                            in_bracket = true;
+                        } else if !(start.is_ascii_alphabetic() || matches!(start, '.' | '_')) {
+                            break 'label_loop;
                         }
 
+                        // Labels continue with ASCII alphanumeric characters, _, or $
+                        for c in chars {
+                            // Inside a template format arg, any character is permitted for the puproses of label detection
+                            // because we assume that it can be replaced with some other valid label string later.
+                            // `options(raw)` asm blocks cannot have format args, so they are excluded from this special case.
+                            if !raw && in_bracket {
+                                if c == '{' {
+                                    // Nested brackets are not allowed in format args, this cannot be a label.
+                                    break 'label_loop;
+                                }
+
+                                if c == '}' {
+                                    // The end of the format arg.
+                                    in_bracket = false;
+                                }
+                            } else if !raw && c == '{' {
+                                // Start of a format arg.
+                                in_bracket = true;
+                            } else {
+                                if !(c.is_ascii_alphanumeric() || matches!(c, '_' | '$')) {
+                                    // The potential label had an invalid character inside it, it cannot be a label.
+                                    break 'label_loop;
+                                }
+                            }
+                        }
+
+                        // If all characters passed the label checks, this is likely a label.
+                        found_labels.push(possible_label);
                         start_idx = idx + 1;
                     }
                 }

--- a/tests/ui/asm/named-asm-labels.rs
+++ b/tests/ui/asm/named-asm-labels.rs
@@ -120,6 +120,27 @@ fn main() {
         // is there an example that is valid x86 for this test?
         asm!(":bbb nop");
 
+        // non-ascii characters are not allowed in labels, so should not trigger the lint
+        asm!("Ù: nop");
+        asm!("testÙ: nop");
+        asm!("_Ù_: nop");
+
+        // Format arguments should be conservatively assumed to be valid characters in labels
+        // Would emit `test_rax:` or similar
+        #[allow(asm_sub_register)]
+        {
+            asm!("test_{}: nop", in(reg) 10); //~ ERROR avoid using named labels
+        }
+        asm!("test_{}: nop", const 10); //~ ERROR avoid using named labels
+        asm!("test_{}: nop", sym main); //~ ERROR avoid using named labels
+        asm!("{}_test: nop", const 10); //~ ERROR avoid using named labels
+        asm!("test_{}_test: nop", const 10); //~ ERROR avoid using named labels
+        asm!("{}: nop", const 10); //~ ERROR avoid using named labels
+
+        asm!("{uwu}: nop", uwu = const 10); //~ ERROR avoid using named labels
+        asm!("{0}: nop", const 10); //~ ERROR avoid using named labels
+        asm!("{1}: nop", "/* {0} */", const 10, const 20); //~ ERROR avoid using named labels
+
         // Test include_str in asm
         asm!(include_str!("named-asm-labels.s")); //~ ERROR avoid using named labels
 

--- a/tests/ui/asm/named-asm-labels.stderr
+++ b/tests/ui/asm/named-asm-labels.stderr
@@ -245,7 +245,88 @@ LL |             ab: nop // ab: does foo
    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
 
 error: avoid using named labels in inline assembly
-  --> $DIR/named-asm-labels.rs:124:14
+  --> $DIR/named-asm-labels.rs:132:19
+   |
+LL |             asm!("test_{}: nop", in(reg) 10);
+   |                   ^^^^^^^
+   |
+   = help: only local labels of the form `<number>:` should be used in inline asm
+   = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
+
+error: avoid using named labels in inline assembly
+  --> $DIR/named-asm-labels.rs:134:15
+   |
+LL |         asm!("test_{}: nop", const 10);
+   |               ^^^^^^^
+   |
+   = help: only local labels of the form `<number>:` should be used in inline asm
+   = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
+
+error: avoid using named labels in inline assembly
+  --> $DIR/named-asm-labels.rs:135:15
+   |
+LL |         asm!("test_{}: nop", sym main);
+   |               ^^^^^^^
+   |
+   = help: only local labels of the form `<number>:` should be used in inline asm
+   = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
+
+error: avoid using named labels in inline assembly
+  --> $DIR/named-asm-labels.rs:136:15
+   |
+LL |         asm!("{}_test: nop", const 10);
+   |               ^^^^^^^
+   |
+   = help: only local labels of the form `<number>:` should be used in inline asm
+   = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
+
+error: avoid using named labels in inline assembly
+  --> $DIR/named-asm-labels.rs:137:15
+   |
+LL |         asm!("test_{}_test: nop", const 10);
+   |               ^^^^^^^^^^^^
+   |
+   = help: only local labels of the form `<number>:` should be used in inline asm
+   = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
+
+error: avoid using named labels in inline assembly
+  --> $DIR/named-asm-labels.rs:138:15
+   |
+LL |         asm!("{}: nop", const 10);
+   |               ^^
+   |
+   = help: only local labels of the form `<number>:` should be used in inline asm
+   = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
+
+error: avoid using named labels in inline assembly
+  --> $DIR/named-asm-labels.rs:140:15
+   |
+LL |         asm!("{uwu}: nop", uwu = const 10);
+   |               ^^^^^
+   |
+   = help: only local labels of the form `<number>:` should be used in inline asm
+   = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
+
+error: avoid using named labels in inline assembly
+  --> $DIR/named-asm-labels.rs:141:15
+   |
+LL |         asm!("{0}: nop", const 10);
+   |               ^^^
+   |
+   = help: only local labels of the form `<number>:` should be used in inline asm
+   = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
+
+error: avoid using named labels in inline assembly
+  --> $DIR/named-asm-labels.rs:142:15
+   |
+LL |         asm!("{1}: nop", "/* {0} */", const 10, const 20);
+   |               ^^^
+   |
+   = help: only local labels of the form `<number>:` should be used in inline asm
+   = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
+
+error: avoid using named labels in inline assembly
+  --> $DIR/named-asm-labels.rs:145:14
    |
 LL |         asm!(include_str!("named-asm-labels.s"));
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
@@ -254,7 +335,7 @@ LL |         asm!(include_str!("named-asm-labels.s"));
    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
 
 warning: avoid using named labels in inline assembly
-  --> $DIR/named-asm-labels.rs:134:19
+  --> $DIR/named-asm-labels.rs:155:19
    |
 LL |             asm!("warned: nop");
    |                   ^^^^^^
@@ -262,13 +343,13 @@ LL |             asm!("warned: nop");
    = help: only local labels of the form `<number>:` should be used in inline asm
    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
 note: the lint level is defined here
-  --> $DIR/named-asm-labels.rs:132:16
+  --> $DIR/named-asm-labels.rs:153:16
    |
 LL |         #[warn(named_asm_labels)]
    |                ^^^^^^^^^^^^^^^^
 
 error: avoid using named labels in inline assembly
-  --> $DIR/named-asm-labels.rs:143:20
+  --> $DIR/named-asm-labels.rs:164:20
    |
 LL |     unsafe { asm!(".Lfoo: mov rax, {}; ret;", "nop", const 1, options(noreturn)) }
    |                    ^^^^^
@@ -277,7 +358,7 @@ LL |     unsafe { asm!(".Lfoo: mov rax, {}; ret;", "nop", const 1, options(noret
    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
 
 error: avoid using named labels in inline assembly
-  --> $DIR/named-asm-labels.rs:149:20
+  --> $DIR/named-asm-labels.rs:170:20
    |
 LL |     unsafe { asm!(".Lbar: mov rax, {}; ret;", "nop", const 1, options(noreturn)) }
    |                    ^^^^^
@@ -286,7 +367,7 @@ LL |     unsafe { asm!(".Lbar: mov rax, {}; ret;", "nop", const 1, options(noret
    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
 
 error: avoid using named labels in inline assembly
-  --> $DIR/named-asm-labels.rs:157:20
+  --> $DIR/named-asm-labels.rs:178:20
    |
 LL |     unsafe { asm!(".Laaa: nop; ret;", options(noreturn)) }
    |                    ^^^^^
@@ -295,7 +376,7 @@ LL |     unsafe { asm!(".Laaa: nop; ret;", options(noreturn)) }
    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
 
 error: avoid using named labels in inline assembly
-  --> $DIR/named-asm-labels.rs:167:24
+  --> $DIR/named-asm-labels.rs:188:24
    |
 LL |         unsafe { asm!(".Lbbb: nop; ret;", options(noreturn)) }
    |                        ^^^^^
@@ -304,7 +385,7 @@ LL |         unsafe { asm!(".Lbbb: nop; ret;", options(noreturn)) }
    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
 
 error: avoid using named labels in inline assembly
-  --> $DIR/named-asm-labels.rs:176:15
+  --> $DIR/named-asm-labels.rs:197:15
    |
 LL |         asm!("closure1: nop");
    |               ^^^^^^^^
@@ -313,7 +394,7 @@ LL |         asm!("closure1: nop");
    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
 
 error: avoid using named labels in inline assembly
-  --> $DIR/named-asm-labels.rs:180:15
+  --> $DIR/named-asm-labels.rs:201:15
    |
 LL |         asm!("closure2: nop");
    |               ^^^^^^^^
@@ -322,7 +403,7 @@ LL |         asm!("closure2: nop");
    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
 
 error: avoid using named labels in inline assembly
-  --> $DIR/named-asm-labels.rs:190:19
+  --> $DIR/named-asm-labels.rs:211:19
    |
 LL |             asm!("closure3: nop");
    |                   ^^^^^^^^
@@ -330,5 +411,5 @@ LL |             asm!("closure3: nop");
    = help: only local labels of the form `<number>:` should be used in inline asm
    = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
 
-error: aborting due to 35 previous errors; 1 warning emitted
+error: aborting due to 44 previous errors; 1 warning emitted
 

